cli/cmd/plugin/plugin.go

```diff
@@ -0,0 +1,234 @@
+	BuildSHA string `json:"buildSHA" yaml:"buildSHA"`
+
+	// Digest is the SHA256 hash of the plugin binary.
+	Digest string `json:"digest" yaml:"digest"`
+
+	// Command group for the plugin.
+	Group CmdGroup `json:"group" yaml:"group"`
+
+	// DocURL for the plugin.
+	DocURL string `json:"docURL,omitempty" yaml:"docURL,omitempty"`
+
+	// Hidden tells whether the plugin should be hidden from the help command.
+	Hidden bool `json:"hidden,omitempty" yaml:"hidden,omitempty"`
+
+	// CompletionType determines how command line completion will be determined.
+	CompletionType PluginCompletionType `json:"completionType" yaml:"completionType"`
+
+	// CompletionArgs contains the valid command line completion values if `CompletionType`
+	// is set to `StaticPluginCompletion`.
+	CompletionArgs []string `json:"completionArgs,omitempty" yaml:"completionArgs,omitempty"`
+
+	// CompletionCommand is the command to call from the plugin to retrieve a list of
+	// valid completion nouns when `CompletionType` is set to `DynamicPluginCompletion`.
+	CompletionCommand string `json:"completionCmd,omitempty" yaml:"completionCmd,omitempty"`
+
+	// Aliases are other text strings used to call this command
+	Aliases []string `json:"aliases,omitempty" yaml:"aliases,omitempty"`
+
+	// InstallationPath is a relative installation path for a plugin binary.
+	// E.g., cluster/v0.3.2@sha256:...
+	InstallationPath string `json:"installationPath"`
+
+	// Discovery is the name of the discovery from where
+	// this plugin is discovered.
+	Discovery string `json:"discovery"`
+
+	// Scope is the scope of the plugin. Stand-Alone or Context
+	Scope string `json:"scope"`
+}
+
+// NewPlugin creates an instance of Plugin.
+func NewPlugin(descriptor *PluginDescriptor) (*Plugin, error) {
+	p := &Plugin{
+		Cmd: &cobra.Command{
+			Use:     descriptor.Name,
+			Short:   descriptor.Description,
+			Aliases: descriptor.Aliases,
+		},
+	}
+
+	// TODO(stmcginnis): There is a docs command we may want to add, need to
+	// see if that's necessary.
+	p.Cmd.AddCommand(
+		newDescribeCmd(descriptor.Description),
+		newVersionCmd(descriptor.Version),
+		newInfoCmd(descriptor),
+		newLintCmd(),
+	)
+	return p, nil
+}
+
+// NewTestFor creates a plugin descriptor for a test plugin.
+func NewTestFor(pluginName string) *PluginDescriptor {
+	return &PluginDescriptor{
+		Name:        fmt.Sprintf("%s-test", pluginName),
+		Description: fmt.Sprintf("test for %s", pluginName),
+		Version:     "v0.0.1",
+		BuildSHA:    SHA,
+		Group:       TestCmdGroup,
+		Aliases:     []string{fmt.Sprintf("%s-alias", pluginName)},
+	}
+}
+
+func newDescribeCmd(description string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:    "describe",
+		Short:  "Describes the plugin",
+		Hidden: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			fmt.Println(description)
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func newVersionCmd(version string) *cobra.Command {
```

Comment:
I was kind of hoping we could drop the auto-version behavior, since at least 2 of our plugins override this. However, it's probably best to keep this as a "just in case" thing to ensure compatibility.

Reply:
Yeah, I'd be fine dropping this. Or we can refactor things so we have some sort of callback or better way to display the version information that we want.

If the plan is to eventually get back to using some sort of plugin API out of the `tanzu-framework` repo though, we might be better off just leaving this as is and overriding where needed so nothing changes too much when that switch happens.

Reply:
agreed

Reply:
Cool - I'm fine with leaving it to make sure we don't diverge.